### PR TITLE
prevent `Compilation.GetSymbolsWithName() from returning null entries

### PR DESCRIPTION
**Pull request by brettfo**
_Fixes internal bug 173790._

We've received crash reports when searching within Solution Explorer.  Unfortunately the dumps weren't properly saved, but we do have a call stack which reports an NRE.

```
   at Microsoft.CodeAnalysis.FindSymbols.SymbolFinder.<FilterByCriteria>d__25.MoveNext() in F:\Builds\5443\DevDiv\Roslyn-Stabilization-Signed-Release\src\Open\src\Workspaces\Core\Portable\FindSymbols\SymbolFinder_Declarations.cs:line 336
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection) in f:\dd\ndp\clr\src\BCL\system\collections\generic\list.cs:line 743
   at Microsoft.CodeAnalysis.FindSymbols.SymbolFinder.<FindSourceDeclarationsAsync>d__24.MoveNext() in F:\Builds\5443\DevDiv\Roslyn-Stabilization-Signed-Release\src\Open\src\Workspaces\Core\Portable\FindSymbols\SymbolFinder_Declarations.cs:line 327
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) in f:\dd\ndp\clr\src\BCL\system\runtime\compilerservices\TaskAwaiter.cs:line 176
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) in f:\dd\ndp\clr\src\BCL\system\runtime\compilerservices\TaskAwaiter.cs:line 156
   at Microsoft.VisualStudio.LanguageServices.Implementation.Progression.SearchGraphQuery.<FindNavigableSourceSymbolsAsync>d__6.MoveNext() in F:\Builds\5443\DevDiv\Roslyn-Stabilization-Signed-Release\src\Open\src\VisualStudio\Core\Def\Implementation\Progression\GraphQueries\SearchGraphQuery.cs:line 120
```

The offending line is [here](https://github.com/dotnet/roslyn/blob/Visual-Studio-2015-Update-1/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs#L336) and the only two methods which pass their results into `FilterByCriteria` are [`Compilation.GetSymbolsWithName()`]() and [`Find()`](https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Declarations.cs#L303).  I traced the calls in `Find()` and never found anything that could return `null` so I went into `Compilation.GetSymbolsWithName()` and eventually found that in both C# and VB, there is a `GetSymbol()` method that can return null which could cause a null item to be added to the result set.  Refactoring this method to be `TryGetSymbol()` seemed heavy-handed since there would still have to be a check before adding the symbol to the result set.  While auditing the code I also found that `GetSpineSymbol()` also calls `GetSymbol()` and could potentially end up returning `null` [here](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L3016) so I guarded the callsites to `GetSpineSymbol()`, too, with the notable exceptions of [here](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L2942) and [here](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L2950) because the subsequent calls go through `GetSymbol()` which is able to handle the `null` parameter.

@dotnet/roslyn-ide

_Update:_
After digging around some user feedback comments I was finally able to nail down a repro for VB and I've verified that without this fix the test fails with the same NRE seen originally.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -2984,13 +2992,16 @@ public IEnumerable<ISymbol> GetSymbolsWithName(Func<string, bool> predicate, Sym
                 spine.Add(current);
 
                 var container = GetSpineSymbol(spine);
-                foreach (var member in container.GetMembers())
```
Comment (brettfo) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Not sure why this diff looks so painful, but I just wrapped the `foreach` in an `if` block.

**Comment by brettfo** [comment]
test prtest/win/dbg/unit32 please
// [build timeout](http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/2773/)

**Comment by brettfo** [comment]
Ping @dotnet/roslyn-ide again for review.

**Comment by Pilchie** [comment]
I see the null returns in `CSharpCompilation.SymbolSearcher.GetCachedSymbol`, but do we actually expect to hit that?  What does it mean to have a declaration for which there is no symbol?

**Comment by brettfo** [comment]
@Pilchie: All results from calling `GetCachedSymbol()` in both C# and VB are either guarded against `null`/`Nothing` or came from `GetSymbol()` directly, so those `null`/`Nothing`s are covered.  As to your question about what it means to have a symbol-less declaration, I have no idea so I'm going to deflect to @heejaechang whom I believe to have originally written this, but I suspect it has to do with broken/incomplete code scenarios where the symbol is an error type.

**Comment by heejaechang** [comment]
I think guard us from crashing is one, but I am wondering how that can ever happen. it would be nice we can get a dump to see what actually happened.

**Comment by heejaechang** [comment]
:+1:

**Comment by brettfo** [comment]
@Pilchie, to maybe better answer your question of `how could this ever be `null`?', it can happen if the [`AddCache()`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Compilation/CSharpCompilation.cs,2967023f5c79d602) method doesn't match a container symbol's type.  Looking through the type hierarchy, this could happen in error cases like `MissingNamespaceSymbol` and `ErrorTypeSymbol`, so I think `null` is a valid scenario if there were binding errors.

**Comment by heejaechang** [comment]
@brettfo are you saying in error case,

private NamespaceOrTypeSymbol GetSymbol(NamespaceOrTypeSymbol container, MergedNamespaceOrTypeDeclaration declaration)
            {
                if (container == null)
                {
                    return _compilation.GlobalNamespace;
                }
 
                if (declaration.Kind == DeclarationKind.Namespace)
                {
                    AddCache(container.GetMembers(declaration.Name).OfType<NamespaceOrTypeSymbol>());
                }
                else
                {
                    AddCache(container.GetTypeMembers(declaration.Name));
                }
 
                return GetCachedSymbol(declaration);
            }

container.GetXXXX(declaration.Name) doesnt return error symbol or

private void AddCache(IEnumerable<NamespaceOrTypeSymbol> symbols)
            {
                foreach (var symbol in symbols)
                {
                    var mergedNamespace = symbol as MergedNamespaceSymbol;
                    if (mergedNamespace != null)
                    {
                        _cache[mergedNamespace.ConstituentNamespaces.OfType<SourceNamespaceSymbol>().First().MergedDeclaration] = symbol;
                        continue;
                    }
 
                    var sourceNamespace = symbol as SourceNamespaceSymbol;
                    if (sourceNamespace != null)
                    {
                        _cache[sourceNamespace.MergedDeclaration] = sourceNamespace;
                        continue;
                    }
 
                    var sourceType = symbol as SourceMemberContainerTypeSymbol;
                    if (sourceType != null)
                    {
                        _cache[sourceType.MergedDeclaration] = sourceType;
                    }
                }
            }

symbol is there, but this method is missing for error symbol case?

**Comment by brettfo** [comment]
I'm saying that if any of the items returned by `container.GetMembers()` or `container.GetTypeMembers()` in `GetSymbol()` are either `MissingNamespaceSymbol` or `ErrorTypeSymbol`, then `AddCache()` won't match them which could cause `GetCachedSymbol()` to return null and I think that's a valid scenario which means it's valid that `GetSymbol()` could return `null`.

**Comment by heejaechang** [comment]
@brettfo I understand, but why question is can we also add support for those error symbols in AddCache if possible.

just add like

var missingNamespace = symbol as MissingNamespaceSymbol;
if (missingNamespace != null)
{
_cache[... declaration .. ] = missingNamespace;
}

and etc.

**Comment by brettfo** [comment]
I see what your saying, now.

We can certainly add those types to `AddCache()` (and I'll do that shortly).  If I do add the additional checks, should I keep the `null` checks at the call sites to cover anything else that would have missed the categorization, or should `AddCache()` (throw|Debug.Fail|crash|etc.) if it reaches the end of the loop?

**Comment by heejaechang** [comment]
@brettfo I think null check is still good.

**Comment by brettfo** [comment]
I've found a VB repro and added a test, but I was unable to find a C# repro but I still think the C# fix is valuable.  In short, the problem arose when `GetSymbol()` was called in VB, `GetCachedSymbol()` returned `Nothing` so `AddCache()` was called, but `container.GetMembers()` returned an empty set so nothing was added to the cache which made the next call to `GetCachedSymbol()` return `Nothing` again.  Since this is an error case, there's nothing reasonable we can do in `AddCache()` like I originally hoped, but the existing fix prevents the result set from containing `null`/`Nothing` which I think is the correct fix.

**Comment by balajikris** [comment]
LGTM.

**Comment by brettfo** [comment]
Adding a note that this has also appeared as internal bug 199411.

**Pull request merged**